eth/consensus/clique/snapshot_manager.py

```diff
@@ -59,27 +58,28 @@ class SnapshotManager:
     """
 
     logger = get_extended_debug_logger('eth.consensus.clique.snapshot_manager.SnapshotManager')
+    # TODO: How do we get around this? It is way too slow without cached snapshots.
+    _snapshots = lru.LRU(IN_MEMORY_SNAPSHOTS)
```

Comment:
@carver I've hit another road block here with the stateless `ConsensusAPI` concept. I'm sorry I didn't notice this earlier. Here it goes:

So, we were able to get rid of the `HeaderCache` because that could be replaced by going from `validate_seal(header)` to `validate_seal_extension(header, parents)`. I should also note that other clients don't necessarily have a header cache in their clique implementation and that this was just me trying to make things work with just `validate_seal` that only takes a single header.

But the snapshot cache is something different. Every Clique implementation out there has this (afaik) and I believe it is at the very heart of clique so I'm not sure how we would get around it.

It basically means: If we don't cache snapshots then with every single header that we validate, we will have to load all previous header from the database until we hit the most recent checkpoint header (every 30000 header for Görli). As you can imagine, this is super super slow.

As a band-aid fix I moved this to the class level but that obviously isn't an acceptable solution. I can't help myself but think we have to move at least parts this back to the chain level unless you have other clever ideas that I might not be seeing.

I want to note that if you feel like playing with this (you can run it with https://github.com/ethereum/trinity/pull/1196 as simple as `trinity --goerli --disable-tx-pool`) then I'm happy if you take this in any direction that gets us closer to a merge. I'm not trying to offload this to you, just saying that if you have ideas then you should not feel obligated to discuss them first if you think you could just take a stab on it and finish it up. I basically just want to see this wrapped up and out of my way :sweat_smile:

Reply:
Ah, okay. Silly question: what if we persisted every snapshot?

An alternative that comes to mind is having the chain pass in a volatile data store. Much like how it passes in the DB, but with the expectation that everything in the volatile store will be dropped on shutdown. (Also, that the VMs need to manage cross-fork incompatibility across the volatile store themselves, like they need to do with the DB already)

Reply:
>what if we persisted every snapshot?

Multiple problems I think:

1. Lot's of overhead in snapshot creation
2. Still one db lookup per block to get the snapshot
3. Lots of wasted disk space that we need to clean up again

>Much like how it passes in the DB, but with the expectation that everything in the volatile store will be dropped on shutdown.

Yes, dropping the data on shutdown won't be a problem I think even dropping them between forks wouldn't be a problem at all. It would just be one header after the fork that causes us to suck in 30000 headers from the db until we are back to normal.

So, yeah, that would work but I think it can not be just some predefined storage because e.g. in this case we want to setup an LRU cache for the snapshots and in other cases we might needs some different kind of storage or eviction polity.

But actually, this brought me to some idea and I took a stab at it and it seems to work. Here's a rough outline and then I continue with some inline comments.

Instead of the `ConsensusAPI` taking in the db, it will take in a `ConsensusContextAPI`.

```python
class ConsensusAPI(ABC):

    @abstractmethod
    def __init__(self, context: ConsensusContextAPI) -> None:
        """
        Initialize the consensus api.
        """
        ...
```

We set the `consensus_context_class` on the chain. And just like the `ChainDB` the `consensus_context` is set *once* when the chain is initialized.

```python
class Chain(BaseChain):
    ...
    chaindb_class: Type[ChainDatabaseAPI] = ChainDB
    consensus_context_class: Type[ConsensusContextAPI] = ConsensusContext

    def __init__(self, base_db: AtomicDatabaseAPI) -> None:
       ...
        self.chaindb = self.get_chaindb_class()(base_db)
        self.consensus_context = self.consensus_context_class(self.chaindb.db)
        self.headerdb = HeaderDB(base_db)
```

Then later, in `get_vm(at_header)` that context is being used.

```python
    def get_vm(self, at_header: BlockHeaderAPI = None) -> VirtualMachineAPI:
        header = self.ensure_header(at_header)
        vm_class = self.get_vm_class_for_block_number(header.block_number)
        chain_context = ChainContext(self.chain_id)

        return vm_class(
            header=header,
            chaindb=self.chaindb,
            chain_context=chain_context,
            consensus_context=self.consensus_context
        )
```

So, the bottom line is:

- we continue to have VM instances that do only operate on the state that they get passed in
- we have consensus context state that is per-chain rather than per-vm
- we do not have any state leaking (as in: two different chain instances accidentially sharing consensus state).

There's a lot to clean up here but you can take a look at [the most recent commit](https://github.com/ethereum/py-evm/pull/1899/commits/6e8484ba6475433280119167b884c76de8b594fd) which implements this and also another one [on the Trinity side](https://github.com/ethereum/trinity/pull/1196) that takes it on a ride. It works and I think it kinds gives us the best of both worlds.

I have a good feeling with that, hoping you feel the same vibe :sweat_smile:
If you are :+1: on that direction than I can clean up the mess and propose a clean PR.

Reply:
Yeah, this feels like a good direction :+1: :tada: 

I also like the idea that maybe we can enforce the reset of the context across fork boundaries. For example, we might create an independent consensus_context for each distinct VM, so in `get_vm()` instead of just getting `self.consensus_context` we get `self.get_consensus_context(vm_class)` which is generated on demand, but only has one per VM.

PoW could use this too! There is an epoch cache that it doesn't want to regenerate every time, and I think it does something hacky to save them (like keep a global cache :see_no_evil: ). Obviously that change doesn't need to go in here, but a tracking issue would be cool.

Reply:
>PoW could use this too! There is an epoch cache that it doesn't want to regenerate every time, and I think it does something hacky to save them

Ah yes! We can use that to clean this up :)

>I also like the idea that maybe we can enforce the reset of the context across fork boundaries

Yeah, we could do that if we need it but I think we should have a good reason to do that. Like in, the Clique case it would just cause us to take more time to validate the header after each fork transition while not providing any actual benefit. But sure there could be other consensus schemes that would need a different context per fork.

Ok, I'm gonna start getting this into shape now :)

Reply:
Tracking issue: https://github.com/ethereum/py-evm/issues/1900